src/org/ods/orchestration/util/DocumentHistory.groovy

```diff
@@ -0,0 +1,332 @@
+        }
+        def content = repo.loadFile(fileName)
+        try {
+            content.collect { Map entry ->
+                if (! entry.entryId) {
+                    throw new IllegalArgumentException('EntryId cannot be empty')
+                }
+                if (! entry.projectVersion) {
+                    throw new IllegalArgumentException('projectVersion cannot be empty')
+                }
+                new DocumentHistoryEntry(
+                    entry,
+                    entry.entryId,
+                    entry.projectVersion,
+                    entry.previousProjectVersion,
+                    entry.rational
+                )
+            }
+        } catch (Exception e) {
+            throw new IllegalArgumentException('Unable to load saved document history for file ' +
+                "'${fileName}': ${e.message}", e)
+        }
+    }
+
+    String saveDocHistoryData(ProjectDataBitbucketRepository repository) {
+        repository.save(this.data, this.getSavedDocumentName())
+    }
+
+    List<DocumentHistoryEntry> getDocHistoryEntries() {
+        this.data
+    }
+
+    List<Map> getDocGenFormat() {
+        def issueTypes = JiraDataItem.TYPES - JiraDataItem.TYPE_DOCS
+        def transformEntry =  { DocumentHistoryEntry e ->
+            if (e.getEntryId() == 1L) {
+                return [
+                    entryId: e.getEntryId(),
+                    rational: e.getRational(),
+                    ]
+            }
+            def formatedIssues = issueTypes.collect { type ->
+                def issues = e.getOrDefault(type, [])
+                if (issues.isEmpty()) {
+                    return null
+                }
+                def changed = issues.findAll { it.action == CHANGE }.clone()
+                    .collect { [key: it.key, predecessors: it.predecessors.join(", ")] }
+
+                [ type: type,
+                  (ADDED): SortUtil.sortIssuesByKey(issues.findAll { it.action == ADD }),
+                  (CHANGED): SortUtil.sortIssuesByKey(changed),
+                  (DELETED): SortUtil.sortIssuesByKey(issues.findAll { it.action == DELETE }),
+                ]
+            }.findAll { it }
+
+            return [entryId: e.getEntryId(),
+                    rational: e.getRational(),
+                    issueType: formatedIssues + computeDocChaptersOfDocument(e)
+            ]
+        }
+        sortDocHistories(this.data).collect { transformEntry(it) }
+    }
+
+    @SuppressWarnings(['UseCollectMany'])
+    protected List<String> getDocumentKeys() {
+        def result = this.data
+            .collect { e ->
+                e.findAll { JiraDataItem.TYPES.contains(it.key) }
+                    .collect { type, actions -> actions.collect { it.key } }
+            }.flatten()
+        if (result) {
+            return result
+        }
+        return []
+    }
+
+    protected Map computeDocChaptersOfDocument(DocumentHistoryEntry entry) {
```

Comment:
to improve encapsulation all the following `protected method in this class could be set to `private`. Inheritance is not needed atm as there is no child class of `DocumentHistory`